src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+// Asynchronous memory module
+
+package dinocpu
+
+import chisel3._
+import chisel3.util._
+
+import chisel3.experimental.ChiselEnum
+
+import chisel3.util.experimental.loadMemoryFromFile
+import firrtl.annotations.MemoryLoadFileType
+
+// A Bundle used for representing a memory access by instruction memory or data memory.
+class Request extends Bundle {  
+  val address   = UInt(32.W)
+  val writedata = UInt(32.W)
+
+  // Encodes the operation requested by this access
+  //  0 -> nop
+  //  1 -> read
+  //  2 -> write
+  val operation = UInt(2.W)
+}
+
+// A Bundle used for temporarily storing the necessary information for a write in the data memory accessor.
+class PartialWrite extends Bundle {
+  val address   = UInt(32.W)
+  val writedata = UInt(32.W)
+  val maskmode  = UInt(2.W)
+
+  // Used to determine whether the stored write is actually a valid write.
+  val valid     = Bool() 
+}
+
+/** 
+ * The generic interface for communication between the IMem/DMemAccess modules and the backing memory.
+ *
+ * Input:  request, the ready/valid interface for a MemAccess module to issue Requests to. Memory
+ *         will only accept a request when both request.valid (the MemAccess is supplying valid data)
+ *         and request.ready (the memory is idling for a request) are high.
+ *
+ * Output: response, the valid interface for the data outputted by memory if it was requested to read. 
+ *         the bits in response.bits should only be treated as valid data when response.valid is high.
+ */
+class AsyncMemIO extends Bundle {
+  val request  = Flipped(Decoupled (new Request))
+  val response = Valid (UInt (32.W))
```

Comment:
BTW, this is really cool that you found this `Valid` interface. I like it a lot, and I was totally unaware!